README.md

```diff
@@ -1,55 +1,510 @@
+can easily see if it contains the 20 elements we expect.
+
+```ocaml
+ let test () =
+   let q = create () in
+   let domainA = Domain.spawn (fun () -> work 1 q) in
+   let domainB = Domain.spawn (fun () -> work 2 q) in
+   Domain.join domainA;
+   Domain.join domainB;
+   (List.length !q, !q)
+```
+
+Let's try it :
+
+```ocaml
+# test ()
+- : int * int list =
+(20, [2; 2; 2; 2; 2; 2; 2; 2; 2; 2; 1; 1; 1; 1; 1; 1; 1; 1; 1; 1])
+```
+
+Everything seems fine, right? Except, it is not running in parallel, as we can
+see from the consecutive `2` (pushed by `domainB`) and `1` pushed by `domainA`.
+This is because spawning a domain takes way more time than executing `work`, so
+`domainA` is long finished before `domainB` is even spawned. One way to overpass
+this issue is to increase the amount of work done in `work` (for example, by
+pushing more elements). Another way is to make sure the domains wait for each
+other before beginning their workload.
+
+We use a basic
+[barrier implementation](https://github.com/ocaml-multicore/saturn/blob/main/test/barrier/barrier.mli)
+to do that. Each domain will now wait for the other to reach the barrier before
+beginning to push.
+
+```ocaml
+let work_par id barrier q =
+  Barrier.await barrier;
+  for i = 0 to 9 do
+    push q id
+  done
+```
+
+The `test` function is now:
+
+```ocaml
+let test_par () =
+  let barrier = Barrier.create 2 in
+  let q = create () in
+  let domainA = Domain.spawn (fun () -> work_par 1 barrier q) in
+  let domainB = Domain.spawn (fun () -> work_par 2 barrier q) in
+  Domain.join domainA;
+  Domain.join domainB;
+  (List.length !q, !q)
+```
+
+Let's run it:
+
+```ocaml
+# test_par ();;
+- : int * int list =
+(18, [2; 1; 2; 1; 2; 1; 2; 1; 2; 1; 2; 2; 1; 1; 1; 1; 2; 1])
+```
+
+Now, the `1` and the `2` are interleaved: domains are running in parallel. The
+resulting queue however only contains `18` elements whereas `20` were pushed.
+This is because we not only have a
+[race condition](https://en.wikipedia.org/wiki/Race_condition) here but also
+because `push` is a non-atomic operation. It requires first to read the content
+of the queue (`!q`) then to write in it (`q := ...`). So, for example, when two
+domains try to push in parallel into an empty queue the following sequence can
+happen:
+
+- domain A reads the queue : it is empty
+- domain B reads the queue : it is still empty
+- domain A pushes `1` on the empty queue it has read before
+- domain B pushes `2` on the empty queue it has read before.
+
+This sequence results in a queue containing only one element of value `2`. The
+element pushed by A is lost because B did not see it.
+
+This is a very common issue in parallel programming. To prevent it, functions
+need to be atomically consistent (aka linearisable), meaning they must have a
+linearisation point at which they appear to occur instantly. Such function can
+be written with different technics, including:
```

Comment:
typo: techniques (and maybe "Such function**s**")

Reply:
Sorry for the "typo" reviews, I missed the warning not to report those.. But otherwise the new readme is a great read, well done! :)

Reply:
That's alright, I should have removed it  ! Thanks for the review !